File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? mergeKLists(List<ListNode?> lists) {
    if(lists.length == 0) return null;
    ListNode? res = lists[0];
    for(int i=1;i<lists.length;i++){
        res = merge(res, lists[i]);
    }
    return res;
  }
  ListNode? merge(ListNode? list1, ListNode? list2){
    if(list2 == null) return list1;
    ListNode? res = new ListNode();
    ListNode? cur = res;
    while(list1 != null || list2 != null){
        if(list2 == null || (list1 != null && list1.val <= list2.val)){
            cur!.next = list1;
            list1 = list1!.next;
        }else {
            cur!.next = list2;
            list2 = list2.next;
        }
        cur = cur.next;
    }
    return res.next;
  }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == 0) return nullptr;
        else if(lists.size() == 1) return lists[0];
        vector<ListNode*> temp;
        for(int i=0;i<lists.size()-1;i+=2){
            temp.push_back(mergerTow(lists[i], lists[i+1]));
        }
        if(lists.size() % 2 == 1){
            temp.push_back(lists[lists.size()-1]);
        }
        return mergeKLists(temp);
    }

    ListNode* mergerTow(ListNode* node1, ListNode* node2){
        ListNode* res = new ListNode();
        ListNode* cur = res;
        while(node1 && node2){
            if(node1->val < node2->val){
                cur->next = new ListNode(node1->val);
                node1 = node1->next;
            }else {
                cur->next = new ListNode(node2->val);
                node2 = node2->next;
            }
            cur = cur->next;
        }
        cur->next = node1 ? node1 : node2;
        return res->next;
    }
};